File: hellogo/pkg/chess/optimized/board.go
// Package optimized provides optimized data structures for chess engine implementation.
package optimized

import "strconv"

// Board represents a chess board with piece placement and move history.
type Board struct {
	pieces      []int
	playedMoves *IntList
}

// NewBoard creates a new board from a FEN string.
func NewBoard(fen string) *Board {
	b := &Board{
		playedMoves: NewIntList(),
		pieces:      make([]int, 120),
	}
	b.fillBlockers()

	rank := 7
	file := 0
	for _, c := range fen {
		if c >= '1' && c <= '8' {
			count := int(c - '0')
			if file+count > 8 {
				panic("Too many pieces on rank " + strconv.Itoa(rank+1))
			}
			for i := 0; i < count; i++ {
				b.pieces[21+rank*10+file] = None
				file++
			}
		} else if c == '/' {
			if file != 8 {
				panic("Invalid FEN: incomplete rank " + strconv.Itoa(rank+1))
			}
			rank--
			file = 0
		} else {
			piece := FromCode(c)
			if piece == None {
				panic("Invalid FEN: unknown piece " + string(c))
			}
			b.pieces[21+rank*10+file] = piece
			file++
		}
	}
	return b
}

// NewBoardCopy creates a copy of an existing board.
func NewBoardCopy(original *Board) *Board {
	pieces := make([]int, len(original.pieces))
	copy(pieces, original.pieces)
	return &Board{
		pieces:      pieces,
		playedMoves: NewIntListFrom(original.playedMoves),
	}
}

// fillBlockers initializes the board edges with blocker pieces.
func (b *Board) fillBlockers() {
	// Fill top and bottom borders
	for i := 0; i < 20; i++ {
		b.pieces[i] = Blocker
		b.pieces[100+i] = Blocker
	}
	// Fill left and right borders
	for i := 2; i < 10; i++ {
		startRank := i * 10
		b.pieces[startRank] = Blocker
		b.pieces[startRank+9] = Blocker
	}
}

// GetSquare converts a UCI square string to an internal square index.
func GetSquare(uciSquare string) int {
	if len(uciSquare) != 2 {
		panic("Invalid UCI square: " + uciSquare)
	}
	file := int(uciSquare[0] - 'a')
	rank := int(uciSquare[1] - '1')
	return 21 + 10*rank + file
}

// GetRank returns the rank (0-7) of a square.
func GetRank(square int) int {
	return (square - 21) / 10
}

// GetUCI converts an internal square index to a UCI string.
func GetUCI(square int) string {
	square -= 21
	file := byte(square%10 + 'a')
	rank := strconv.Itoa(square/10 + 1)
	return string([]byte{file, rank[0]})
}

// GetPiece returns the piece at a given UCI square.
func (b *Board) GetPiece(uciSquare string) int {
	return b.pieces[GetSquare(uciSquare)]
}

// getPiece returns the piece at a given internal square index.
func (b *Board) getPiece(square int) int {
	return b.pieces[square]
}

// MakeMove applies a move to the board.
func (b *Board) MakeMove(move int) {
	from := From(move)
	to := To(move)

	// Move the piece
	b.pieces[to] = b.pieces[from]
	b.pieces[from] = None

	// Record the move with the captured piece
	b.playedMoves.Add(move)
}

// UnmakeMove undoes the last move made on the board.
func (b *Board) UnmakeMove() {
	// Get the last move and remove it from history
	move := b.playedMoves.RemoveLast()
	from := From(move)
	to := To(move)

	// Restore the moved piece
	b.pieces[from] = b.pieces[to]

	// Restore the captured piece (if any)
	b.pieces[to] = Capture(move)
}

// String returns a string representation of the board.
func (b *Board) String() string {
	var result []byte
	for rank := 7; rank >= 0; rank-- {
		result = append(result, byte('1'+rank), ' ')
		for file := 0; file < 8; file++ {
			square := 21 + rank*10 + file
			result = append(result, byte(GetCode(b.pieces[square])))
		}
		result = append(result, '\n')
	}
	result = append(result, "  abcdefgh"...)
	return string(result)
}

File: hellojava/src/main/java/com/fathzer/chess/optimized/Board.java
package com.fathzer.chess.optimized;

import static com.fathzer.chess.optimized.IntMove.*;
import static com.fathzer.chess.optimized.Piece.*;

import java.util.Arrays;

/** A tiny chess board
 * WARNING: This is a very basic implementation. It is not a chess move generator.
 * It lacks important features like king safety, en passant, castling, promotion.
 */
public class Board {
    private final int[] pieces;
	public Board(Board copy) {
		this.pieces = copy.pieces.clone();
	}

    public Board(String fen) {
        this.pieces = new int[120];
        fillBlockers();
		int rank = 7;
		int file = 0;
		for (char c : fen.toCharArray()) {
			if (Character.isDigit(c)) {
				final int count = Character.getNumericValue(c);
				if (count>8-file) {
					throw new IllegalArgumentException("Invalid FEN: too many pieces on rank " + rank);
				}
				file += count;
			} else if (c == '/') {
				if (file!=8) {
					throw new IllegalArgumentException("Invalid FEN: missing files on rank " + rank);
				}
				if (rank == 0) {
					throw new IllegalArgumentException("Invalid FEN: too many ranks");
				}
				file = 0;
				rank--;
			} else {
				final int piece = Piece.fromCode(c);
				if (piece == Piece.NONE) {
					throw new IllegalArgumentException("Invalid FEN: unknown piece " + c);
				}
				this.pieces[21+rank*10+file] = piece;
				file++;
			}
		}
		if (file!=8) {
			throw new IllegalArgumentException("Invalid FEN: missing files on rank " + rank);
		}
		if (rank!=0) {
			throw new IllegalArgumentException("Invalid FEN: missing ranks");
		}
    }
    
    private void fillBlockers() {
    	Arrays.fill(pieces, 0, 20, BLOCKER);
    	Arrays.fill(pieces, 100, 120, BLOCKER);
    	for (int i=2; i<10; i++) {
			final int startRank = i*10;
    		pieces[startRank] = BLOCKER;
    		pieces[startRank+9] = BLOCKER;
    	}
    }
	
	static int getSquare(String uciSquare) {
		if (uciSquare.length() != 2) {
			throw new IllegalArgumentException("Invalid UCI square: " + uciSquare);
		}
		final char file = uciSquare.charAt(0);
		final char rank = uciSquare.charAt(1);
		if (rank < '1' || rank > '8' || file < 'a' || file > 'h') {
			throw new IllegalArgumentException("Invalid UCI square: " + uciSquare);
		}
		return 21+ 10*(Integer.parseInt(uciSquare.substring(1)) - 1) + (file - 'a');
	}
	
	static int getRank(int square) {
		return (square-21)/10;
	}
	
	static String getUCI(int square) {
		square -= 21;
		return (char)(square%10 + 'a') + "" +(square/10 + 1);
	}

	public int getPiece(String uciSquare) {
		return pieces[getSquare(uciSquare)];
	}

	int getPiece(int square) {
		return pieces[square];
	}

	public IntList getMoves(IntList moves, boolean white) {
		moves.clear();
		for (int square = 20; square < 100; square++) {
			int piece = getPiece(square);
			if (piece != NONE && piece!=BLOCKER && isWhite(piece)==white) {
				MoveGenerators.get(piece).build(moves, this, square);
			}
		}
		return moves;
	}

	public void makeMove(int move) {
	    int from = from(move);
	    int to = to(move);
	    
	    // Move the piece
	    pieces[to] = pieces[from];
	    pieces[from] = NONE; // Empty the source square
	}

	public void unmakeMove(int move) {
	    int from = from(move);
	    int to = to(move);
	    
	    // Restore the moved piece
	    pieces[from] = pieces[to];
	    
	    // Restore the captured piece (if any)
	    pieces[to] = capture(move);
	}

	@Override
	public String toString() {
		final StringBuilder b = new StringBuilder();
		for (char rank = '8'; rank >= '1'; rank--) {
			b.append(rank);
			b.append(' ');
			for (char file = 'a'; file <= 'h'; file++) {
				int piece = getPiece(file+""+rank);
				b.append(getCode(piece));
			}
			b.append('\n');
		}
		b.append("  ");
		for (char file = 'a'; file < 'h' ; file++) {
			b.append(file);
		}
		return b.toString();
	}
}
